run/opencl/gost12256hash_kernel.cl

```diff
@@ -0,0 +1,302 @@
+#else
+	cnt = len;
+#endif
+	GOST34112012Update(&ctx, result.BYTES, cnt, loc_buf);
+
+	/* Take the binary representation of the length of the key and for every
+	   1 add the alternate sum, for every 0 the key. */
+	for (cnt = len; cnt > 0; cnt >>= 1)
+		if ((cnt & 1) != 0)
+			GOST34112012Update(&ctx, result.BYTES, BINARY_SIZE, loc_buf);
+		else
+			GOST34112012Update(&ctx, key, len, loc_buf);
+
+	/* Create intermediate result. */
+	GOST34112012Final(&ctx, &result, loc_buf);
+
+	/* Start computation of P byte sequence. */
+	GOST34112012Init(&alt_ctx, 256);
+
+	/* For every character in the password add the entire password. */
+	for (cnt = 0; cnt < len; ++cnt)
+		GOST34112012Update(&alt_ctx, key, len, loc_buf);
+
+	/* Finish the digest. */
+	GOST34112012Final(&alt_ctx, &temp_result, loc_buf);
+
+	/* Create byte sequence P. */
+	cp = p_bytes;
+#if PLAINTEXT_LENGTH > BINARY_SIZE
+	for (cnt = len; cnt > BINARY_SIZE; cnt -= BINARY_SIZE) {
+		memcpy_pp(cp, &temp_result, BINARY_SIZE);
+		cp += BINARY_SIZE;
+	}
+#else
+	cnt = len;
+#endif
+	memcpy_pp(cp, &temp_result, cnt);
+
+	/* Start computation of S byte sequence. */
+	GOST34112012Init(&alt_ctx, 256);
+
+	/* repeat the following 16+A[0] times, where A[0] represents the
+	   first byte in digest A interpreted as an 8-bit uvalue */
+	for (cnt = 0; cnt < 16 + result.BYTES[0]; ++cnt)
+		GOST34112012Update(&alt_ctx, salt, saltlen, loc_buf);
+
+	/* Finish the digest. */
+	GOST34112012Final(&alt_ctx, &temp_result, loc_buf);
+
+	/* Create byte sequence S. */
+	cp = s_bytes;
+#if SALT_LENGTH > BINARY_SIZE
+	for (cnt = saltlen; cnt > BINARY_SIZE; cnt -= BINARY_SIZE) {
+		memcpy_pp(cp, &temp_result, BINARY_SIZE);
+		cp += BINARY_SIZE;
+	}
+#else
+	cnt = saltlen;
+#endif
+	memcpy_pp(cp, &temp_result, cnt);
+
+	/* Here's everything we need for the loop kernel */
+	memcpy256(&(out[gid]), &(result));
+	memcpy_pg(state[gid].p_bytes, p_bytes, len);
+	memcpy_pg(state[gid].s_bytes, s_bytes, saltlen);
+}
+
+__kernel void gost12256loop(__global inbuf *in,
+                            MAYBE_CONSTANT saltstruct *ssalt,
+                            __global statebuf *state,
+                            __local localbuf *loc_buf,
+                            __global uint256_u *out)
+{
+	GOST34112012Context ctx;
+	uint256_u result;
+	uint gid = get_global_id(0);
+	uchar p_bytes[PLAINTEXT_LENGTH];
+	uchar s_bytes[SALT_LENGTH];
+
+	uint cnt;
+	uint saltlen = ssalt->len;
+	uint len = in[gid].len;
+
+	memcpy256(&(result), &(out[gid]));
+	memcpy_gp(p_bytes, state[gid].p_bytes, len);
+	memcpy_gp(s_bytes, state[gid].s_bytes, saltlen);
+
+#if STREEBOG_LOCAL_AX
+	uint ls = get_local_size(0);
+	uint lid = get_local_id(0);
+
+	for (uint i = lid; i < 256; i += ls) {
+		for (uint j = 0; j < 8; j++)
+			loc_buf->Ax[j][i] = Ax[j][i];
+	}
+	barrier(CLK_LOCAL_MEM_FENCE);
+#endif
+
+	/* Repeatedly run the collected hash value through Streebog to burn CPU cycles.  */
+#pragma unroll HASH_LOOPS
```

Comment:
@magnumripper Maybe we should simply drop these `#pragma unroll HASH_LOOPS` lines from here? Were they meant to be included in the pushed code? There's even a comment in `gost12512hash_kernel.cl` that says:

```
 * Without unroll: Build time: 45.125 s, binary size 2542391, 10402 c/s
 * Pragma unroll:  Build time: 21 min 14.788 s, binary size 11502092, 10051 c/s
 * Manual unroll:  Build time: 4 min 26.142 s, binary size 5961490, 9959 c/s
```

So are you saying you had better build times and better runtime speeds without the unroll, yet kept the unroll? Weird.

Reply:
Testing on various ancient systems, this full unroll results in various "out of resources" errors. Reducing the unroll to 42 also gives such errors, but quicker. Removing the unroll makes the kernel build and invocation succeed, but the the formats commonly fail self-test on `cmp_all`.

Reply:
```c
/*
 * Without unroll: Build time: 45.125 s, binary size 2542391, 10402 c/s
 * Pragma unroll:  Build time: 21 min 14.788 s, binary size 11502092, 10051 c/s
 * Manual unroll:  Build time: 4 min 26.142 s, binary size 5961490, 9959 c/s
 */
#define STREEBOG_UNROLL         0
#define STREEBOG_MANUAL_UNROLL  0
#include "opencl_streebog.h"
```
Those comments are about unrolling within opencl_streebog.h (controlled by the macros that are indeed set to 0 before sourcing it). 

The `#pragma unroll HASH_LOOPS` in gost12512hash_kernel.cl is a different beast. I assume I got better results with it on whatever nvidia I tested with at the time.

Reply:
Also unroll 21 could be reasonable (leaves only the trivial `& 1` checks in the loop) and it even passes my build/run tests (where 42 fails), only failing self-test (wrong compute as usual for these ancient systems). Build time and code size e.g. for `gost94crypt-opencl` with old Catalyst targeting Tahiti is 11 seconds and 1 MB without unroll, 110 seconds and 5 MB with unroll 21.

All of these code sizes are too large. I doubt that even runtime `% 3` and `% 7` checks make much of a difference when code doesn't fit in cache anyway. And I don't know why it's so large without unrolling?

Anyway, instead of unrolling we could use one of the tricks we already use elsewhere. Like maintaining counter(s) that are reset to 0 upon reaching 3, 7, 21, or 42 (and use an offsets map like `md5crypt_long_fmt.c` does). Crazy unrolling is really not the way.